[SPARK-8882][Streaming]Add a new Receiver scheduling mechanism

zsxwing opened the pull request:
The design doc: https://docs.google.com/document/d/1ZsoRvHjpISPrDmSjsGzuSu8UjwgbtmoCTzmhgTurHJw/edit?usp=sharing

zsxwing committed:
[SPARK-5681][Streaming] Use a lock to eliminate the race condition when stopping receivers and registering receivers happen at the same time #6294

zsxwing committed:
Add a new Receiver scheduling mechanism

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36742 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36742/consoleFull) for   PR 7276 at commit [`d9a3e72`](https://github.com/apache/spark/commit/d9a3e720f6f6bce93e6b5f88b7c45a37866cf63f).

SparkQA commented:
[Test build #36742 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36742/console) for   PR 7276 at commit [`d9a3e72`](https://github.com/apache/spark/commit/d9a3e720f6f6bce93e6b5f88b7c45a37866cf63f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

zsxwing committed:
Add JobWaiter.toFuture to avoid blocking threads

zsxwing committed:
Add unit tests for LoadBalanceReceiverSchedulerImplSuite

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36797 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36797/consoleFull) for   PR 7276 at commit [`27acd45`](https://github.com/apache/spark/commit/27acd45d345ef8e364682f03dbd8c3e9eef60feb).

zsxwing committed:
Add a test for Receiver.restart

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36798 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36798/consoleFull) for   PR 7276 at commit [`ca6fe35`](https://github.com/apache/spark/commit/ca6fe359fcbd752eaec8b1e713f1c6cffb90b751).

SparkQA commented:
[Test build #36797 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36797/console) for   PR 7276 at commit [`27acd45`](https://github.com/apache/spark/commit/27acd45d345ef8e364682f03dbd8c3e9eef60feb).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
**[Test build #36798 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36798/console)**     for PR 7276 at commit [`ca6fe35`](https://github.com/apache/spark/commit/ca6fe359fcbd752eaec8b1e713f1c6cffb90b751)     after a configured wait of `175m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

zsxwing commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36868 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36868/consoleFull) for   PR 7276 at commit [`ca6fe35`](https://github.com/apache/spark/commit/ca6fe359fcbd752eaec8b1e713f1c6cffb90b751).

SparkQA commented:
**[Test build #36868 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36868/console)**     for PR 7276 at commit [`ca6fe35`](https://github.com/apache/spark/commit/ca6fe359fcbd752eaec8b1e713f1c6cffb90b751)     after a configured wait of `175m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

zsxwing committed:
Use tryFailure to support calling jobFailed multiple times

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36919 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36919/consoleFull) for   PR 7276 at commit [`2c86a9e`](https://github.com/apache/spark/commit/2c86a9e0ff4dd11ea0680cda45bb4f64c4e01aff).

SparkQA commented:
[Test build #36919 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36919/console) for   PR 7276 at commit [`2c86a9e`](https://github.com/apache/spark/commit/2c86a9e0ff4dd11ea0680cda45bb4f64c4e01aff).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

tdas commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -1860,6 +1860,30 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
+   * Submit a job for execution and return a FutureJob holding the result. Return a Future for
+   * monitoring the job success or failure event.
+   */
+  private[spark] def submitAsyncJob[T, U, R](
```
This probably should be marked experimental, since submitJob is experimental.

@JoshRosen Pinging you to review this part of the PR that touches SparkContext.

JoshRosen replied:
One high-level question, which perhaps has been addressed elsewhere: why can't we use the existing `submitJob` method that returns a `SimpleFutureAction`?

zsxwing replied:
BTW, I use the name `submitAsyncJob` because if using `submitJob`, the type inference cannot work well and many places using the previous `submitJob` cannot be compiled.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/receiver/ReceiverSupervisor.scala:
```diff
@@ -64,6 +64,8 @@ private[streaming] abstract class ReceiverSupervisor(
   /** State of the receiver */
   @volatile private[streaming] var receiverState = Initialized
 
+  val host = Utils.localHostName()
```
Does this need to be public?

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/receiver/ReceiverSupervisor.scala:
```diff
@@ -161,12 +163,20 @@ private[streaming] abstract class ReceiverSupervisor(
     }(futureExecutionContext)
   }
 
+  /** Reschedule this receiver and return a candidate executor list */
```
Using the term `reschedule` is confusing here. I think what you want is essentially to get a set of acceptable locations from the driver. Wouldne a better name be `getAllowedLocations`?

zsxwing committed:
Make 'host' protected; rescheduleReceiver -> getAllowedLocations

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37090 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37090/consoleFull) for   PR 7276 at commit [`28d1bee`](https://github.com/apache/spark/commit/28d1bee9c1afbc23be650d21ef53c946c7c49298).

SparkQA commented:
**[Test build #37090 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37090/console)**     for PR 7276 at commit [`28d1bee`](https://github.com/apache/spark/commit/28d1bee9c1afbc23be650d21ef53c946c7c49298)     after a configured wait of `175m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverScheduler.scala:
```diff
@@ -0,0 +1,87 @@
+      preferredLocation: Option[String],
+      receiverTrackingInfoMap: Map[Int, ReceiverTrackingInfo],
+      executors: Seq[String]): Seq[String]
+}
```
Incorrect indentation.

tdas replied:
Ignore that.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverScheduler.scala:
```diff
@@ -0,0 +1,87 @@
+private[streaming] trait ReceiverScheduler {
+
+  /**
+   * Return a candidate executor list to run the receiver. If the list is empty, the caller can run
```
Doesnt it return a list of executors?

tdas commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -1860,6 +1860,30 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
+      processPartition: (TaskContext, Iterator[T]) => U,
+      resultHandler: (Int, U) => Unit,
+      resultFunc: => R): Future[Unit] =
+  {
```
nit: Shouldnt this be on the previous line.

tdas commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -1860,6 +1860,30 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
+      allowLocal = false,
+      resultHandler,
+      localProperties.get)
+    waiter.toFuture
```
I dont see an unit test for this in the PR.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/receiver/ReceiverSupervisor.scala:
```diff
@@ -161,12 +163,20 @@ private[streaming] abstract class ReceiverSupervisor(
     }(futureExecutionContext)
   }
 
+  /** Reschedule this receiver and return a candidate executor list */
```
Please update the scala docs to match the name.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverScheduler.scala:
```diff
@@ -0,0 +1,87 @@
+
+/**
+ * A ReceiverScheduler trying to balance executors' load.
+ */
```
Can give a detailed description of algorithm that is used here. How the weights and stuff work? Its not immediately obvious from the code, so I am not sure how to verify the correctness of the algorithm.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverScheduler.scala:
```diff
@@ -0,0 +1,87 @@
+          val scheduledLocations = receiverTrackingInfo.scheduledLocations.get
+          // The probability that a scheduled receiver will run in an executor is
+          // 1.0 / scheduledLocations.size
+          scheduledLocations.map(location => (location, 1.0 / scheduledLocations.size))
```
nit: Keep the tuple creation code consistent, either `a -> b` or `(a, b)`, in these lines.

JoshRosen commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -1860,6 +1860,30 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
   }
 
   /**
+   * Submit a job for execution and return a FutureJob holding the result. Return a Future for
```
These two sentences are slightly redundant, I think.

JoshRosen commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverScheduler.scala:
```diff
@@ -0,0 +1,87 @@
+          scheduledLocations.map(location => (location, 1.0 / scheduledLocations.size))
+        case ReceiverState.ACTIVE => Seq(receiverTrackingInfo.runningLocation.get -> 1.0)
+      }
+    }.groupBy(_._1).mapValues(_.map(_._2).sum) // Sum weights for each executor
```
Ignore this comment if this code is not performance-critical, but I wonder whether writing imperative code that updates a mutable hashmap might be more efficient here than materializing all of the values for each group. If we don't go the imperative route, maybe there's still a more efficient functional way to handle this by doing an aggregate-by-key instead of group-by followed by aggregate.  Disregard this comment if perf. isn't a factor here, though.

tdas replied:
perf really isnt a factor. The max cardinality that I can imagine for any of these is 100s.

JoshRosen commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala:
```diff
@@ -75,6 +89,19 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
+
+  @volatile private var stopping = false
+
+  // Track receivers' status for scheduling
```
This is a super-minor style point, but if you write these one-liners as Scaladoc then they'll show up in the quick documentation popups in IntelliJ.  I don't feel super strongly about this and we haven't been consistent about this throughout Spark, but just thought I'd comment since this has been on my mind lately.

JoshRosen commented:
My high-level comment / question: why not use the existing `submitJob` that returns a SimpleFutureAction rather than defining your own asynchronous version of that method?

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala:
```diff
@@ -343,24 +378,61 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
+              }
+            val future = ssc.sparkContext.submitAsyncJob[Receiver[_], Unit, Unit](
+              receiverRDD, startReceiver, (_, _) => Unit, ())
+            // TODO Refactor JobWaiter to avoid creating a new Thread here. Otherwise, it's not
```
This TODO needs to be removed.

zsxwing commented:
> My high-level comment / question: why not use the existing submitJob that returns a SimpleFutureAction rather than defining your own asynchronous version of that method?

The problem of `SimpleFutureAction` is its `onComplete` calls `awaitResult`. So if using `SimpleFutureAction`, each Job needs to block one thread. It means  each running `Receiver` needs one thread in the driver. It may exhaust threads.
```
  override def onComplete[U](func: (Try[T]) => U)(implicit executor: ExecutionContext) {
    executor.execute(new Runnable {
      override def run() {
        func(awaitResult())
      }
    })
  }
```

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala:
```diff
@@ -227,6 +255,8 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
       case DeregisterReceiver(streamId, message, error) =>
         deregisterReceiver(streamId, message, error)
         context.reply(true)
+      case ScheduleReceiver(streamId) =>
```
I think this messaging should be done differently to align the messaging style with the other messages. It does not make sense for the receiver to fetch all the allowed locations that the receiver is allowed to start. Rather it should be 
`ShouldStartReceiver`, which returns true or false, and accordingly the receiver is started by the `ReceiverSupervisor`.

tdas replied:
Or if we want to maintain the naming scheme (which uses `Register` instead `Start`), it may be better to name the new one `ValidateLocation(streamId, host)`

tdas replied:
Actually, we can merge the change you did in #6294, where `ReceiverSupervisor.onReceiverStart()` (that is, `RegisterReceiver` message) is called before the receiver is started, and the receiver is only started if `RegisterReceiver` message returns `true`. This mechanism can be used by the `ReceiverTracker` to prevent the receiver from starting on unwanted executors - when `RegisterReceiver` is received by the tracker, it will return `!stopping && scheduleReceiver(receiverId).contains(host))`. 

This will solve the race condition as well as check for location without introducing additional messaging... isnt it?

zsxwing replied:
Sounds great. One potential issue is if we reject a receiver using `!scheduleReceiver(receiverId).contains(host))` when receiving `RegisterReceiver`, we may keep restarting a receiver if the scheduled executors keep busy, such as running some long running jobs.

My current implementation only rejects a receiver for a mismatch location when restarting it.

tdas replied:
Good point. How about this. 
The scheduler returns the following. 
- if there are >= 3, zero-weight options, return all of them
- else return 3 best options.

Before submitting the job, run the scheduler and use those options to launch the job. When the RegisterReceiver is called, you again run the scheduler to get updated options and verify whether the host where the task was actually launched is still in the best options. That should reduce the likelihood of the above condition happening by a lot. 

How does that sound?

zsxwing replied:
Sounds good. I will update it.

zsxwing committed:
Fix comments and code style

tdas commented:
2 major high-level comments that need significant code changes before we proceed further.
- Messaging can be optimized, see other comments 
- The ReceiverLauncher is becoming quite complex, is it going to be too complex to move it to a separate file?

zsxwing committed:
Add comments for the scheduling approach

zsxwing committed:
Remove submitAsyncJob and revert JobWaiter

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

zsxwing committed:
Revert SparkContext.scala

zsxwing commented:
I have reverted the changes to core. So this PR won't pass the tests until #7385 is merged. I will update this PR then.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37242 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37242/consoleFull) for   PR 7276 at commit [`3b87e4a`](https://github.com/apache/spark/commit/3b87e4ad87dbcfd498adf93883d48caab601c0ed).

SparkQA commented:
[Test build #37242 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37242/console) for   PR 7276 at commit [`3b87e4a`](https://github.com/apache/spark/commit/3b87e4ad87dbcfd498adf93883d48caab601c0ed).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverScheduler.scala:
```diff
@@ -0,0 +1,108 @@
+    scheduledLocations: Option[Seq[String]],
+    runningLocation: Option[String])
+
+private[streaming] trait ReceiverScheduler {
```
Another high level point. Now there are three things. ReceiverTracker, ReceiverLauncher and ReceiverScheduler. Its super confusing to reason what does what. I think I am inclined change the ReceiverScheduler name to ReceiverSchedulingPolicy, so that it clear that its just a policy and not an "active" module at the level of ReceiverTracker and ReceiverScheduling policy.

zsxwing committed:
ReceiverScheduler -> ReceiverSchedulingPolicy

zsxwing committed:
Merge branch 'squash-pr-6294' into receiver-scheduling

zsxwing committed:
Remove the ScheduleReceiver message because we can refuse it when receiving RegisterReceiver

zsxwing committed:
Update tha scheduling algorithm to avoid to keep restarting Receiver

zsxwing commented:
> The ReceiverLauncher is becoming quite complex, is it going to be too complex to move it to a separate file?

How about doing it in a follow-up PR? Moving it to another file would make this PR hard to review.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37383 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37383/consoleFull) for   PR 7276 at commit [`5fee132`](https://github.com/apache/spark/commit/5fee1324be4289e73d847b6a8c1a0ef0cb28555b).

SparkQA commented:
[Test build #37383 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37383/console) for   PR 7276 at commit [`5fee132`](https://github.com/apache/spark/commit/5fee1324be4289e73d847b6a8c1a0ef0cb28555b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

tdas commented:
Okay, sounds good.

On Wednesday, July 15, 2015, Shixiong Zhu <notifications@github.com> wrote:

> The ReceiverLauncher is becoming quite complex, is it going to be too
> complex to move it to a separate file?
>
> How about doing it in a follow-up PR? Moving it to another file would make
> this PR hard to review.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/7276#issuecomment-121685090>.
>


-- 
Sent from Gmail Mobile

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/receiver/ReceiverSupervisorImpl.scala:
```diff
@@ -182,4 +182,5 @@ private[streaming] class ReceiverSupervisorImpl(
     logDebug(s"Cleaning up blocks older then $cleanupThreshTime")
     receivedBlockHandler.cleanupOldBlocks(cleanupThreshTime.milliseconds)
   }
+
```
nit: extra line

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverSchedulingPolicy.scala:
```diff
@@ -0,0 +1,110 @@
+
+import org.apache.spark.streaming.scheduler.ReceiverState._
+
+private[streaming] case class ReceiverTrackingInfo(
```
Please provide scala docs for this class.

tdas replied:
This should not be in this file. ReceiverTrackingInfo is better to be defined in the ReceiverTracker, isnt it?

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverSchedulingPolicy.scala:
```diff
@@ -0,0 +1,110 @@
+ * </ol>
+ *
+ */
+private[streaming] class LoadBalanceReceiverSchedulingPolicyImpl extends ReceiverSchedulingPolicy {
```
LoadBalanceReceiverSchedulingPolicyImpl --> LoadBalance*d*ReceiverSchedulingPolicy

Why add impl? Either use `ReceiverSchedulingPolicyImpl` or `LoadBalancedReceiverSchedulingPolicy` I am fine either ways. In face, the first one might be better since there is only one implementation

tdas replied:
For that matter, why even define the interface? Do envision having other policies in the near future?

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverSchedulingPolicy.scala:
```diff
@@ -0,0 +1,110 @@
+    locations ++= preferredLocation
+
+    val executorWeights = receiverTrackingInfoMap.filter { case (id, _) =>
+      // Ignore the receiver to be scheduled. It may be still running.
```
What does this mean "It may be still running"? Can you elaborate that case, when and how that can happen?

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverSchedulingPolicy.scala:
```diff
@@ -0,0 +1,110 @@
+          val scheduledLocations = receiverTrackingInfo.scheduledLocations.get
+          // The probability that a scheduled receiver will run in an executor is
+          // 1.0 / scheduledLocations.size
+          scheduledLocations.map(location => location -> 1.0 / scheduledLocations.size)
```
put `1.0 / scheduledLocations.size` in parenthesis, becomes easier to read.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala:
```diff
@@ -71,13 +82,57 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
+  private val receiverTrackingInfosLock = new AnyRef
+
+  /** Check if tracker has been marked for starting */
+  private def isTrackerStarted(): Boolean = trackerStateLock.synchronized {
```
nit: Please move these helper methods lower in the class after ` hasUnallocatedBlocks`

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala:
```diff
@@ -371,4 +490,59 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
+    }
+
+  private def updateReceiverScheduledLocations(
+      receiverId: Int, scheduledLocations: Seq[String]): Unit =
```
I am not sure whether there is a formatting policy about it but I prefer always using `{` and `}` in multiline methods.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala:
```diff
@@ -343,24 +424,62 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
-      running = false
-      logInfo("All of the receivers have been terminated")
+      for (receiver <- receivers) {
+        submitJobThread.execute(new Runnable {
```
Please make this runnable a separate class. Its too nested here.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala:
```diff
@@ -343,24 +424,62 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
+              return
+            }
+
+            val self = this
```
This block of code needs comments.

tdas commented:
Since #6294 has been merged, please update this PR. A summary of suggestions made offline. 
- Lets try to move all the "starting of receivers" into the event loop of ReceiverTracker. So all the receiverTrackignInfo will be updated and accessed in a lock free manner from the event loop. 
- I encourage you to get rid of ReceiverLauncher if its easy enough.
- The scheduling policy and the scheduled location still needs to be discussed further out (what we discussed offline still has drawbacks).

zsxwing committed:
Merge branch 'master' into receiver-scheduling

zsxwing committed:
Reorganize ReceiverTracker to use an event loop for lock free

1. Add fields into ReceiverTrackingInfo to contain all necessary information for ReceiverInfo. When sending ReceiverInfo to StreamingListener, create ReceiverInfo from ReceiverTrackingInfo directly.
2. Move methods that use ReceiverTrackingInfo into ReceiverTrackerEndpoint so that we can maintain all ReceiverTrackingInfos in an event loop.
3. Move ReceiverTrackingInfo to a separate file ReceiverTrackingInfo.scala.
4. Address some code style issues

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

zsxwing commented:
I reorganized ReceiverTracker to use an event loop, including:

1. Add fields into ReceiverTrackingInfo to contain all necessary information for ReceiverInfo. When sending ReceiverInfo to StreamingListener, create ReceiverInfo from ReceiverTrackingInfo directly.
2. Move methods that use ReceiverTrackingInfo into ReceiverTrackerEndpoint so that we can maintain all ReceiverTrackingInfos in an event loop.
3. Move ReceiverTrackingInfo to a separate file ReceiverTrackingInfo.scala.
4. Address some code style issues

> how does it behave when there are 100 executors, and 100 receivers need to be scheduled in the beginning. ideally there should be 1 receiver per executor

Assume we only have 2 executors and 2 receivers. The current scheduling policy will always return the 2 executors. Then it's possible that 2 receivers run in the same executor. But if returning only 1 executor, we may keep restarting receivers if the location does not match.

zsxwing commented:
@JoshRosen do you know how to get the executor id in a spark job closure?

SparkQA commented:
[Test build #37961 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37961/consoleFull) for   PR 7276 at commit [`f60d021`](https://github.com/apache/spark/commit/f60d0217151cdc72c05671437ef3433bd1c18bba).

zsxwing committed:
Fix unintentional changes

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37964 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37964/consoleFull) for   PR 7276 at commit [`4e639c4`](https://github.com/apache/spark/commit/4e639c4bee03f4cc1dc3c7514ad473e6cb3acabd).

SparkQA commented:
[Test build #37964 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37964/console) for   PR 7276 at commit [`4e639c4`](https://github.com/apache/spark/commit/4e639c4bee03f4cc1dc3c7514ad473e6cb3acabd).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

JoshRosen commented:
@zsxwing, it looks like `SparkEnv` has an `executorId` field that you can use.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala:
```diff
@@ -86,6 +116,12 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
   // This not being null means the tracker has been started and not stopped
   private var endpoint: RpcEndpointRef = null
 
+  private val schedulingPolicy: ReceiverSchedulingPolicy = new ReceiverSchedulingPolicy()
```
nit: `: ReceiverSchedulingPolicy` not needed.

SparkQA commented:
[Test build #37961 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37961/console) for   PR 7276 at commit [`f60d021`](https://github.com/apache/spark/commit/f60d0217151cdc72c05671437ef3433bd1c18bba).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala:
```diff
@@ -170,101 +207,64 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
+    private val submitJobThreadPool = ExecutionContext.fromExecutorService(
+      ThreadUtils.newDaemonCachedThreadPool("submit-job-thead-pool"))
+
     override def receive: PartialFunction[Any, Unit] = {
```
Is there any advantage to separating out the `receive` from the `receiveAndReply`? Isnt it easy enough to not reply in a receiverAndReply?

zsxwing replied:
Actually, `receiverAndReply` needs to create a temp `Actor` to send the message. Since we don't need to reply anything, it's better to use `receive`.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala:
```diff
@@ -274,18 +274,202 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
+      case AllReceiverIds =>
+        context.reply(receiverTrackingInfos.keys.toSeq)
       case StopAllReceivers =>
         assert(isTrackerStopping || isTrackerStopped)
```
Cant these conditions of `stopping || stopped` be simply `!isTrackerStarted()`.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala:
```diff
@@ -274,18 +274,202 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
+      receiverPreferredLocations(receiver.streamId) = receiver.preferredLocation
+      val receiverId = receiver.streamId
+
+      if (isTrackerStopping() || isTrackerStopped()) {
```
Cant these conditions of stopping || stopped be simply !isTrackerStarted().

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala:
```diff
@@ -274,18 +274,202 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
+      // Create the RDD using the scheduledLocations to run the receiver in a Spark job
+      val receiverRDD: RDD[Receiver[_]] =
+        if (scheduledLocations.isEmpty) {
+          ssc.sc.makeRDD(Seq(receiver), 1)
```
nit: Add name to the RDD so that it appears in the UI with useful names like "Receiver 1"

tdas commented:
i took a medium-level pass, not too detailed and overall approach into the event loop looks. However I have some high level comments. 

1. In the code refactoring, why did you move all the `registerReceiver`, `deregisterReceiver`, etc. into the Endpoint class? That moves around a lot of code, making it hard to review. Also it makes the overall structure more complex - some functionality is in the parent ReceiverTracker class and some are in the EndPoint class, hard to reason about what goes where in the future. Instead the earlier structure had the goal of all the functionality is in the ReceiverTracker, and the Endpoint class was just to going to be the event loop that accesses that functionality. I think its more structured if that is maintained.

2. I think the scheduling policy needs to be thought through. It may be in the case we need two steps, one of bulk scheduling in the beginning, which tries to distribute it properly (n recvrs in n execs) by setting preferred locations on all of them. Then for every receiver that needs to be restarted, they are scheduled one at a time. The schedulingPolicy can be changed to take an array of receiverIds to accommodate this, initially all the receiverIds are sent, and the locations of all of them returned; then one at a time. How does that sound?

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala:
```diff
@@ -354,46 +503,22 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
+        receiverJobExitLatch.await()
         logInfo("All of the receivers have been terminated")
       } finally {
         running = false
```
I think the `running` variable can be completely removed. The stop() method can wait on the receiverJobExistLatch directly. This thread does not need to stay blocked.

zsxwing replied:
Since we don't need to block, the thread is not necessary. Then we can also remove `ReceiverLauncher`. Therefore, I removed `ReceiverLauncher` and added a method `launchReceivers` to replace it.

zsxwing commented:
> In the code refactoring, why did you move all the registerReceiver, deregisterReceiver, etc. into the Endpoint class? That moves around a lot of code, making it hard to review. Also it makes the overall structure more complex - some functionality is in the parent ReceiverTracker class and some are in the EndPoint class, hard to reason about what goes where in the future. Instead the earlier structure had the goal of all the functionality is in the ReceiverTracker, and the Endpoint class was just to going to be the event loop that accesses that functionality. I think its more structured if that is maintained.

I moved all methods and fields that are only accessed in the endpoint into the endpoint. I feel it's easy to see they are only accessed in the endpoint. However, it really makes this PR hard to review. Let me move them back to the outer class.

zsxwing committed:
Move methods back to the outer class

zsxwing committed:
Remove "ReceiverLauncher" and move codes to "launchReceivers"

zsxwing committed:
Add receiver RDD name; use '!isTrackerStarted' instead

zsxwing committed:
Schedule all receivers at the same time when launching them

zsxwing committed:
Use hostPort as the receiver location rather than host; fix comments and unit tests

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

zsxwing commented:
After considering carefully, I think if a receiver has a preferredLocation, it means it's better to run the receiver in this preferredLocation if possible. So I don't clear it when restart a receiver.

SparkQA commented:
[Test build #38084 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38084/consoleFull) for   PR 7276 at commit [`8f93c8d`](https://github.com/apache/spark/commit/8f93c8d1aec224a279c7f247d95e708072063d66).

SparkQA commented:
[Test build #38084 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38084/console) for   PR 7276 at commit [`8f93c8d`](https://github.com/apache/spark/commit/8f93c8d1aec224a279c7f247d95e708072063d66).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

zsxwing commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverSchedulingPolicy.scala:
```diff
@@ -0,0 +1,161 @@
+   *
+   * This method is called when a receiver is registering with ReceiverTracker or is restarting.
+   */
+  def rescheduleReceiver(
```
This method was called `scheduleReceiver`. I renamed it to `rescheduleReceiver` to avoid confusing with `scheduleReceivers`.

zsxwing commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala:
```diff
@@ -251,171 +312,238 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
-  }
 
-  /** Check if tracker has been marked for starting */
-  private def isTrackerStarted(): Boolean = trackerState == Started
```
`isTrackerStarted`, `isTrackerStopping` and `isTrackerStopped` is moved avoid ReceiverTrackerEndpoint so that there is no method of `ReceiverTracker` below `ReceiverTrackerEndpoint`.

zsxwing commented:
> @zsxwing, it looks like SparkEnv has an executorId field that you can use.

@JoshRosen Thank you. It works.

SparkQA commented:
[Test build #1161 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1161/consoleFull) for   PR 7276 at commit [`8f93c8d`](https://github.com/apache/spark/commit/8f93c8d1aec224a279c7f247d95e708072063d66).

SparkQA commented:
[Test build #1161 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1161/console) for   PR 7276 at commit [`8f93c8d`](https://github.com/apache/spark/commit/8f93c8d1aec224a279c7f247d95e708072063d66).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #1180 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1180/consoleFull) for   PR 7276 at commit [`8f93c8d`](https://github.com/apache/spark/commit/8f93c8d1aec224a279c7f247d95e708072063d66).

SparkQA commented:
[Test build #1180 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1180/console) for   PR 7276 at commit [`8f93c8d`](https://github.com/apache/spark/commit/8f93c8d1aec224a279c7f247d95e708072063d66).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala:
```diff
@@ -189,13 +238,21 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
+          runningLocation = Some(hostPort),
+          name = Some(name),
+          endpoint = Some(receiverEndpoint)))
+      listenerBus.post(StreamingListenerReceiverStarted(receiverInfo))
```
nit: isnt it cleaner to use the receiverTrackingInfo.toReceiverInfo() here?

zsxwing replied:
Right. Fixed it.

zsxwing committed:
Merge branch 'master' into receiver-scheduling

zsxwing committed:
Use receiverTrackingInfo.toReceiverInfo

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38144 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38144/consoleFull) for   PR 7276 at commit [`05daf9c`](https://github.com/apache/spark/commit/05daf9c829c6c17d1e7ebd4042e687c6627b0b90).

SparkQA commented:
[Test build #38144 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38144/console) for   PR 7276 at commit [`05daf9c`](https://github.com/apache/spark/commit/05daf9c829c6c17d1e7ebd4042e687c6627b0b90).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala:
```diff
@@ -258,171 +319,242 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
   /** RpcEndpoint to receive messages from the receivers. */
   private class ReceiverTrackerEndpoint(override val rpcEnv: RpcEnv) extends ThreadSafeRpcEndpoint {
 
+    // TODO Remove this thread pool after https://github.com/apache/spark/issues/7385 is merged
```
Is there any drawback in implementing in the current way? Is it that with 100 receivers, there will be 100 threads stuck.

zsxwing replied:
> Is it that with 100 receivers, there will be 100 threads stuck.

Right. That's why we need #7385.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala:
```diff
@@ -258,171 +319,242 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
+          receiverId,
+          ReceiverState.SCHEDULED,
+          Some(scheduledLocations),
+          None)
```
nit: What is none for? Could you write it as `xyz = None`? Easier to read.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala:
```diff
@@ -258,171 +319,242 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
+  /**
+   * Get the list of executors excluding driver
+   */
+  private def getExecutors: Seq[String] = {
```
Isnt it easier to return the (host, port) split, so that you dnt have to split it later in the scheduling policy. Either way I am fine, just think about it.

zsxwing replied:
I prefer using `String`. `(host, port)` requires only one `split` in the scheduling policy, but `String` requires to use `host + ":" + port` in several places.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala:
```diff
@@ -203,21 +262,20 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
-          if (error == null || error == "") -1 else ssc.scheduler.clock.getTimeMillis()
-        oldInfo.copy(endpoint = null, active = false, lastErrorMessage = message,
-          lastError = error, lastErrorTime = lastErrorTime)
+        oldInfo.copy(errorInfo = Some(errorInfo))
```
Dont you have to mark the ReceiverState as INACTIVE here?

zsxwing replied:
Set `state` to `INACTIVE`. However, I don't remove `runningLocation` because it will be used to create `ReceiverInfo`.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverSchedulingPolicy.scala:
```diff
@@ -0,0 +1,161 @@
+    }
+
+    val hostToExecutors = executors.groupBy(_.split(":")(0))
+    val locations = new Array[mutable.ArrayBuffer[String]](receivers.length)
```
You can make this `Array.fill(new mutable.ArrayBuffer[String]]`

tdas replied:
Also, `locations` is getting confusing below, because (1) there is `scheduledLocations` later, and (2) in general the name `location` and `executor` have overlapping meaning between `executor` and `location`, which is very confusing.

How about sticking to a convention: `location` as in `preferredLocation` is a host, not an executor. `Executor` is obviously named.

So `locations` in the above lined can be renamed to `scheduledExecutors`, `scheduledLocation` in line 60 can be renamed to `leastScheduledExecutor`.

Also, this convention should be following in the `ReceiverTracker` and `ReceiverTrackingInfo` as well. That is, `ReceiverTrackingInfo` should have `scheduledExecutors` and `runningExecutors`.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverSchedulingPolicy.scala:
```diff
@@ -0,0 +1,161 @@
+    val locations = new Array[mutable.ArrayBuffer[String]](receivers.length)
+    val numReceiversOnExecutor = mutable.HashMap[String, Int]()
+    // Set the initial value to 0
+    executors.foreach(numReceiversOnExecutor(_) = 0)
```
nit: A little easier to read if written as `executors.foreach { e => numReceiversOnExecutor(e) = 1 }

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverSchedulingPolicy.scala:
```diff
@@ -0,0 +1,161 @@
+   * `preferredLocation`s of receivers are not even, we may not be able to schedule them evenly
+   * because we have to respect them.
+   *
+   * This method is called when we start to launch receivers at the first time.
```
Would be good to elaborate the algorithm in the comments. .. things like 
1. First, schedule all the receivers with preferred locations (hosts), evenly among the executors running on those hosts
2. Then, schedule all other receivers evenly among all the executors such that overall distribution over all the receivers is even.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverSchedulingPolicy.scala:
```diff
@@ -0,0 +1,161 @@
+            // 1. This executor is not up. But it may be up later.
+            // 2. This executor is dead, or it's not a host in the cluster.
+            // Currently, simply add host to the scheduled locations
+            locations(i) += host
```
Why add this host? So that if an executor happens to come up, it will get matched?

zsxwing replied:
Yes. That's why I add it.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverSchedulingPolicy.scala:
```diff
@@ -0,0 +1,161 @@
+    val idleExecutors = numReceiversOnExecutor.filter(_._2 == 0).map(_._1)
+    for (executor <- idleExecutors) {
+      // Assign an idle executor to the receiver that has least locations.
+      val scheduledLocations = locations.minBy(_.size)
```
This assigns these extra locations to receivers with preferred locations as well. I dont think we should do that.

zsxwing replied:
But it's possible that we have some idle executors and every receiver has preferred location. In this case, I feel we should also schedule idle executors just in case the preferred executors are busy.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverSchedulingPolicy.scala:
```diff
@@ -0,0 +1,161 @@
+
+    // For those receivers that don't have preferredLocation, make sure we assign at least one
+    // executor to them.
+    for (scheduledLocations <- locations.filter(_.isEmpty)) {
```
This is not correct. What if there are not executors empty as all of them have receivers with preferred locations? You still have to set preferred locations for other receivers as well, even if it mean allocating the same executors once again. The current code will not allocate any location to some of the receivers in this case.

zsxwing replied:
This is selecting the receivers without any scheduled location. And the size of `numReceiversOnExecutor` is same as the size of `executors`, and must not be empty. After this `foreach`, it should allocate a location for all receivers.

tdas replied:
Aaah, okay. Man, I am getting confused between these scheduledLocations and locations and executors. Mind if you do the renaming?

zsxwing replied:
Sure. Renamed as per your suggestion.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverSchedulingPolicy.scala:
```diff
@@ -0,0 +1,161 @@
+    }
+
+    // Always try to schedule to the preferred locations
+    val locations = mutable.Set[String]()
```
As I said above, please use the convention. Less confusing.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverSchedulingPolicy.scala:
```diff
@@ -0,0 +1,161 @@
+      receiverId: Int,
+      preferredLocation: Option[String],
+      receiverTrackingInfoMap: Map[Int, ReceiverTrackingInfo],
+      executors: Seq[String]): Seq[String] = {
```
Might be good make the `3` as a parameter, so that we can add logic later in the ReceiverTracker to increment this number for every failed start.

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverSchedulingPolicySuite.scala:
```diff
@@ -0,0 +1,147 @@
+/**
+ * Dummy receiver implementation
+ */
+private class DummyReceiver(receiverId: Int, host: Option[String] = None)
```
Why move the dummy receiver here? Was it necessary in any way?

zsxwing replied:
`ReceiverTrackerSuite.scala` was removed previously, so I moved it here. After resolving the conflicts, `ReceiverTrackerSuite.scala` is added back. But I forgot to move `DummyReceiver` back.

Already moved it back to `ReceiverTrackerSuite.scala`

zsxwing committed:
Rename: scheduledLocations -> scheduledExecutors; locations -> executors

zsxwing committed:
Move DummyReceiver back to ReceiverTrackerSuite

zsxwing committed:
Fix the code style

zsxwing committed:
Set state to ReceiverState.INACTIVE in deregisterReceiver

zsxwing committed:
Add preferredNumExecutors to rescheduleReceiver

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38202 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38202/consoleFull) for   PR 7276 at commit [`137b257`](https://github.com/apache/spark/commit/137b257fabaa198706528a008394f41ca247782a).

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverSchedulingPolicy.scala:
```diff
@@ -0,0 +1,171 @@
+   */
+  def rescheduleReceiver(
+      receiverId: Int,
+      preferredLocation: Option[String],
```
Might be more consistent with `scheduleReceivers` to pass the whole Receiver object, rather than receiverId and preferredLocatons.

zsxwing replied:
Using `receiverId` and `preferredLocatons` is because we don't store the Receiver object currently. The Receiver object is only available when launching it.

SparkQA commented:
[Test build #38202 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38202/console) for   PR 7276 at commit [`137b257`](https://github.com/apache/spark/commit/137b257fabaa198706528a008394f41ca247782a).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverSchedulingPolicySuite.scala:
```diff
@@ -0,0 +1,130 @@
+    assert(numReceiversOnExecutor === executors.map(_ -> 1).toMap)
+  }
+
+  test("scheduleReceivers: schedule receivers evenly when the preferredLocations are even") {
```
Would be possible to express these test in the nicer way that ReceiverTracker expressed?
`testScheduler(numReceivers = 5, preferredLocation = false, allocation = "0|1|2|3|0")`

zsxwing replied:
Sorry. Forgot to reply this one. I don't want to use an accurate assertion because the scheduling policy should be able to assign receivers to any executor as long as the finally result is even.

E.g., if we have 2 executors and 2 receivers, the scheduling result could be (receiver 0 -> executor 0, receiver 1 -> executor 1), or (receiver 0 -> executor 1, receiver 1 -> executor 0).

tdas replied:
Got it, makes sense.

tdas commented:
I think this PR is close to LGTM. I will wait for #7385 to get merged, then this PR can be updated accordingly.

tdas commented:
I chatted with @pwendell for his opinion on this, and its fine to merge this as is without the #7385. Once that is merged we can always improve the performance later. Not that its going to really hurt performance in the current state, as it will be just a bunch of sleeping blocked threads, there is not constant thread switching.

zsxwing commented:
@tdas Agree. We can do it later.

tdas commented:
Alright, then i am merging this!
